### PR TITLE
Refactor get replication slots

### PR DESCRIPTION
**Pull request by matthbakeredb**
Reduce complexity of single method and allow for documentation of distinct parts.

No functional changes have been introduced.

Split from #2743

### CODE REVIEW

File: patroni/dcs/__init__.py [hunk 1]
```diff
@@ -583,36 +583,63 @@ def use_slots(self) -> bool:
...
+
+        :param my_name: name of this node.
+        :param role: role of this node.
+        :param nofailover: ``True`` if this node is tagged to not be a failover candidate.
+        :param major_version: postgresql major version.
+        :param show_error: if ``True`` report error if any disabled logical slots or conflicting slot names are found.
+
+        :returns: final dictionary of slot names, after merging with permanent slots and performing sanity checks.
+        """
+        slot_members: list[str] = self._get_slot_members(my_name, role)
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py:601]: Should we better do something like 
```suggestion
        slot_members: list[str] = self._get_slot_members(my_name, role) if self.use_slots else []
```

and get rid of `use_slots` checks in the `_get_slot_members()`?

File: patroni/dcs/__init__.py [hunk 2]
```diff
@@ -583,36 +583,63 @@ def use_slots(self) -> bool:
...
             for name in slot_members:
                 slot_conflicts[slot_name_from_member_name(name)].append(name)
             logger.error("Following cluster members share a replication slot name: %s",
-                         "; ".join("{} map to {}".format(", ".join(v), k)
+                         "; ".join(f"{', '.join(v)} map to {k}"
                                    for k, v in slot_conflicts.items() if len(v) > 1))
 
-        # "merge" replication slots for members with permanent_replication_slots
+        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
+            slots, self._get_permanent_slots(role, nofailover), my_name, major_version)
```
Comment (CyberDem0n) [inline patroni/dcs/__init__.py:616]: Similar to https://github.com/zalando/patroni/pull/2746/files#diff-247edd60473c0933e528a1b4621cc612da54b6dc964720719d7623135d312474R601:
```suggestion
        permanent_slots = self._get_permanent_slots(role, nofailover) if self.use_slots else {}
        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
            slots, permanent_slots, my_name, major_version)
```

and get rid of `use_slots` checks in the `_get_permanent_slots()`.

File: patroni/dcs/__init__.py (hunk 1 above)
Comment (matthbakeredb) [reply patroni/dcs/__init__.py]: Nice! Yes that reduces some duplication. Committed.

File: patroni/dcs/__init__.py (hunk 2 above)
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:616]: Applied also in commit.

Comment (CyberDem0n) [reply patroni/dcs/__init__.py:616]: You missed this part:
```python
        permanent_slots = self._get_permanent_slots(role, nofailover) if self.use_slots else {}
```

Comment (matthbakeredb) [reply patroni/dcs/__init__.py:616]: Um, where? I don't see it?

Comment (CyberDem0n) [reply patroni/dcs/__init__.py:616]: Currently it looks like:
```python
        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
            slots, self._get_permanent_slots(role, nofailover), my_name, major_version)
```

While it should be:
```python
        permanent_slots = self._get_permanent_slots(role, nofailover) if self.use_slots else {}
        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
            slots, permanent_slots, my_name, major_version)
```

Comment (matthbakeredb) [reply patroni/dcs/__init__.py]: Oh duh! I was looking for something to remove when I should've been looking for something to add. :roll_eyes: 

Added that now. Sorry!

**Review by CyberDem0n — COMMENTED** [review]
A couple of small issues.
We better stick with List instead of list, because the latter will probably not with with python 3.6, which is still widely used according to pypi download stats.

File: patroni/dcs/__init__.py [hunk 3]
```diff
@@ -583,36 +583,64 @@ def use_slots(self) -> bool:
...
-        slots = {slot_name_from_member_name(name): {'type': 'physical'} for name in slot_members}
+        :param my_name: name of this node.
+        :param role: role of this node.
+        :param nofailover: ``True`` if this node is tagged to not be a failover candidate.
+        :param major_version: postgresql major version.
+        :param show_error: if ``True`` report error if any disabled logical slots or conflicting slot names are found.
+
+        :returns: final dictionary of slot names, after merging with permanent slots and performing sanity checks.
+        """
+        slot_members: list[str] = self._get_slot_members(my_name, role) if self.use_slots else []
```
[the remark this answers is not in the document]
Comment (matthbakeredb) [reply patroni/dcs/__init__.py:601]: Ah yes, I do normally use `List` so don't know why I didn't this time!

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**